File: NyaLauncher.Core/Download/GameDownloadService.cs
using System;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using NyaLauncher.Core.Config;
using NyaLauncher.Core.Launch;
using NyaLauncher.Core.Models;

namespace NyaLauncher.Core.Download;

public enum GameDownloadPhase
{
    Idle,
    Preparing,
    Downloading,
    Completed,
    Failed,
    Cancelled
}

public sealed record GameDownloadSnapshot(
    long Revision,
    long TaskId,
    GameDownloadPhase Phase,
    string VersionId,
    int StageIndex,
    string StageName,
    string Detail,
    double Percentage,
    long CompletedBytes,
    long TotalBytes,
    int CompletedFiles,
    int TotalFiles,
    double BytesPerSecond)
{
    public static GameDownloadSnapshot Idle { get; } = new(
        0,
        0,
        GameDownloadPhase.Idle,
        string.Empty,
        0,
        "尚无下载任务",
        "请在资源下载页选择 Minecraft 版本。",
        0,
        0,
        0,
        0,
        0,
        0);

    public bool HasTask => Phase != GameDownloadPhase.Idle;

    public bool IsActive => Phase is GameDownloadPhase.Preparing or GameDownloadPhase.Downloading;

    public bool IsTerminal => Phase is GameDownloadPhase.Completed or
        GameDownloadPhase.Failed or GameDownloadPhase.Cancelled;
}

/// <summary>
/// 包装 MinecraftVersionInstaller，通过阶段/快照状态机对外发布下载进度。
/// </summary>
public sealed class GameDownloadService
{
    public static readonly string[] StageNames =
    [
        "获取版本元数据",
        "分析下载清单",
        "下载游戏客户端",
        "下载依赖库",
        "下载资源索引",
        "下载游戏资源",
        "完成校验与安装"
    ];

    private readonly object _gate = new();
    private readonly MinecraftVersionInstaller _installer = new();
    private readonly ModLoaderInstaller _modLoaderInstaller = new();
    private GameDownloadSnapshot _current = GameDownloadSnapshot.Idle;
    private CancellationTokenSource? _activeTask;
    private long _revision;
    private long _taskId;

    public GameDownloadSnapshot Current
    {
        get
        {
            lock (_gate)
                return _current;
        }
    }

    public event Action<GameDownloadSnapshot>? Changed;

    public async Task<bool> StartAsync(
        MinecraftVersion version,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(version);
        if (string.IsNullOrWhiteSpace(version.Id) || string.IsNullOrWhiteSpace(version.Url))
            return false;

        CancellationTokenSource taskCancellation;
        long taskId;
        lock (_gate)
        {
            if (_activeTask is not null)
                return false;

            taskCancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
            _activeTask = taskCancellation;
            taskId = ++_taskId;
        }

        Publish(new GameDownloadSnapshot(
            NextRevision(),
            taskId,
            GameDownloadPhase.Preparing,
            version.Id,
            1,
            StageNames[0],
            $"正在准备下载 Minecraft {version.Id}",
            0,
            0,
            0,
            0,
            0,
            0));

        try
        {
            var targetRoot = ResolveTargetMinecraftDirectory();
            var progress = new InlineProgress<MinecraftInstallProgress>(update =>
            {
                if (taskCancellation.IsCancellationRequested || taskId != Volatile.Read(ref _taskId))
                    return;

                Publish(new GameDownloadSnapshot(
                    NextRevision(),
                    taskId,
                    GameDownloadPhase.Downloading,
                    version.Id,
                    update.StageIndex,
                    update.StageName,
                    update.Detail,
                    update.Percentage,
                    update.CompletedBytes,
                    update.TotalBytes,
                    update.CompletedFiles,
                    update.TotalFiles,
                    update.BytesPerSecond));
            });
            await _installer.InstallAsync(
                    version.Id,
                    version.Url,
                    targetRoot,
                    progress,
                    taskCancellation.Token)
                .ConfigureAwait(false);

            // 全局默认隔离开启时，为新版本预建隔离内容目录骨架。
            if (LauncherConfig.DefaultVersionIsolation == true)
            {
                ScaffoldIsolatedContentDirectory(targetRoot, version.Id);
            }

            var sourcePath = LauncherConfig.GameDirectory;
            if (string.IsNullOrWhiteSpace(sourcePath))
            {
                LauncherConfig.SaveGameDirectory(targetRoot);
                sourcePath = targetRoot;
            }

            await GameInstanceStore.RefreshAsync(sourcePath).ConfigureAwait(false);
            GameInstanceStore.Select(version.Id);

            var previous = Current;
            Publish(previous with
            {
                Revision = NextRevision(),
                Phase = GameDownloadPhase.Completed,
                StageIndex = MinecraftVersionInstaller.StageCount,
                StageName = StageNames[^1],
                Detail = $"Minecraft {version.Id} 下载并安装完成",
                Percentage = 100
            });
            return true;
        }
        catch (OperationCanceledException)
        {
            PublishTerminal(
                taskId,
                GameDownloadPhase.Cancelled,
                "下载已取消",
                $"Minecraft {version.Id} 下载已取消");
            return false;
        }
        catch (Exception exception)
        {
            PublishTerminal(
                taskId,
                GameDownloadPhase.Failed,
                "下载失败",
                exception.Message);
            return false;
        }
        finally
        {
            lock (_gate)
            {
                if (ReferenceEquals(_activeTask, taskCancellation))
                {
                    _activeTask.Dispose();
                    _activeTask = null;
                }
            }
        }
    }

    /// <summary>
    /// 以 Mod Loader 模式下载：先确保原版已安装，再叠加安装 Loader。
    /// </summary>
    public async Task<bool> StartModLoaderAsync(
        MinecraftVersion version,
        ModLoaderVersion loader,
        string instanceName,
        bool skipFabricApi = false,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(version);
        ArgumentNullException.ThrowIfNull(loader);
        ArgumentException.ThrowIfNullOrWhiteSpace(instanceName);

        CancellationTokenSource taskCancellation;
        long taskId;
        lock (_gate)
        {
            if (_activeTask is not null)
                return false;

            taskCancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
            _activeTask = taskCancellation;
            taskId = ++_taskId;
        }

        var displayName = $"{loader.Type} {loader.LoaderVersion} (Minecraft {version.Id})";
        Publish(new GameDownloadSnapshot(
            NextRevision(),
            taskId,
            GameDownloadPhase.Preparing,
            instanceName,
            1,
            StageNames[0],
            $"正在准备下载 {displayName}",
            0, 0, 0, 0, 0, 0));

        try
        {
            var targetRoot = ResolveTargetMinecraftDirectory();
            var progress = new InlineProgress<MinecraftInstallProgress>(update =>
            {
                if (taskCancellation.IsCancellationRequested || taskId != Volatile.Read(ref _taskId))
                    return;

                Publish(new GameDownloadSnapshot(
                    NextRevision(),
                    taskId,
                    GameDownloadPhase.Downloading,
                    instanceName,
                    update.StageIndex,
                    update.StageName,
                    update.Detail,
                    update.Percentage,
                    update.CompletedBytes,
                    update.TotalBytes,
                    update.CompletedFiles,
                    update.TotalFiles,
                    update.BytesPerSecond));
            });

            await _modLoaderInstaller.InstallAsync(
                    loader,
                    instanceName,
                    targetRoot,
                    version.Id,
                    progress,
                    taskCancellation.Token)
                .ConfigureAwait(false);

            if (LauncherConfig.DefaultVersionIsolation == true)
            {
                ScaffoldIsolatedContentDirectory(targetRoot, instanceName);
            }

            var sourcePath = LauncherConfig.GameDirectory;
            if (string.IsNullOrWhiteSpace(sourcePath))
            {
                LauncherConfig.SaveGameDirectory(targetRoot);
                sourcePath = targetRoot;
            }

            // Fabric 实例默认随本体一起下载 Fabric API（除非用户在下载选项中勾选跳过）。
            if (loader.Type == ModLoaderType.Fabric && !skipFabricApi)
            {
                await DownloadFabricApiIfNeededAsync(
                        loader,
                        instanceName,
                        targetRoot,
                        sourcePath,
                        version.Id,
                        taskId,
                        taskCancellation.Token)
                    .ConfigureAwait(false);
            }

            await GameInstanceStore.RefreshAsync(sourcePath).ConfigureAwait(false);
            GameInstanceStore.Select(instanceName);

            var previous = Current;
            Publish(previous with
            {
                Revision = NextRevision(),
                Phase = GameDownloadPhase.Completed,
                StageIndex = MinecraftVersionInstaller.StageCount,
                StageName = StageNames[^1],
                Detail = $"{displayName} 安装完成",
                Percentage = 100
            });
            return true;
        }
        catch (OperationCanceledException)
        {
            PublishTerminal(taskId, GameDownloadPhase.Cancelled, "下载已取消", $"{displayName} 下载已取消");
            return false;
        }
        catch (Exception exception)
        {
            PublishTerminal(taskId, GameDownloadPhase.Failed, "下载失败", exception.Message);
            return false;
        }
        finally
        {
            lock (_gate)
            {
                if (ReferenceEquals(_activeTask, taskCancellation))
                {
                    _activeTask.Dispose();
                    _activeTask = null;
                }
            }
        }
    }

    public bool CancelActive()
    {
        lock (_gate)
        {
            if (_activeTask is null)
                return false;
            _activeTask.Cancel();
            return true;
        }
    }

    private static string ResolveTargetMinecraftDirectory()
    {
        var configured = LauncherConfig.GameDirectory ??
                         Environment.GetEnvironmentVariable("NYALAUNCHER_MINECRAFT_DIR") ??
                         MinecraftDirectoryLocator.EnsureDefaultDirectory();
        var fullPath = Path.TrimEndingDirectorySeparator(Path.GetFullPath(configured));
        var directory = new DirectoryInfo(fullPath);
        if (directory.Parent is { } versionsDirectory &&
            string.Equals(
                versionsDirectory.Name,
                "versions",
                StringComparison.OrdinalIgnoreCase) &&
            versionsDirectory.Parent is not null)
        {
            fullPath = versionsDirectory.Parent.FullName;
        }

        Directory.CreateDirectory(fullPath);
        Directory.CreateDirectory(Path.Combine(fullPath, "versions"));
        return fullPath;
    }

    /// <summary>
    /// 在版本隔离启用时，于版本目录下创建标准 Minecraft 内容子文件夹骨架。
    /// 游戏启动时会以该目录作为 <c>--gameDir</c>，提前创建可避免首次运行时目录缺失。
    /// </summary>
    private static void ScaffoldIsolatedContentDirectory(string minecraftRoot, string versionId)
    {
        var versionDirectory = Path.Combine(minecraftRoot, "versions", versionId);
        if (!Directory.Exists(versionDirectory))
            return;

        foreach (var sub in IsolatedContentSubDirectories)
        {
            Directory.CreateDirectory(Path.Combine(versionDirectory, sub));
        }
    }

    private static readonly string[] IsolatedContentSubDirectories =
    [
        "saves",
        "resourcepacks",
        "mods",
        "config",
        "crash-reports",
        "logs",
        "screenshots",
        "shaderpacks",
    ];

    /// <summary>
    /// Modrinth 上的 Fabric API 项目 ID（任何 Fabric 实例都需要的运行库）。
    /// 与 ModDetailDialog 中使用的常量保持一致。
    /// </summary>
    private const string FabricApiProjectId = "P7dR8mSH";

    private const string FabricApiDisplayName = "Fabric API";

    /// <summary>
    /// Fabric 实例安装完成后，默认把 Fabric API 一起下载到该实例的 mods 目录。
    /// 失败不阻断整个安装流程，仅作为警告提示。
    /// </summary>
    private async Task DownloadFabricApiIfNeededAsync(
        ModLoaderVersion loader,
        string instanceName,
        string targetRoot,
        string sourcePath,
        string gameVersion,
        long taskId,
        CancellationToken cancellationToken)
    {
        if (loader.Type != ModLoaderType.Fabric)
            return;

        // 解析该实例实际的内容目录（版本隔离 / 共享布局 / 外部实例均可正确命中）。
        // 全局默认只作兜底传入，不覆盖自动检测（与 GameVersionIsolation.Resolve 语义一致）。
        var layout = GameInstanceLayoutResolver.Resolve(
            targetRoot, sourcePath, instanceName, null, LauncherConfig.DefaultVersionIsolation);
        var modsDir = Path.Combine(layout.ContentDirectory, "mods");
        Directory.CreateDirectory(modsDir);

        try
        {
            Publish(new GameDownloadSnapshot(
                NextRevision(), taskId, GameDownloadPhase.Downloading, instanceName,
                MinecraftVersionInstaller.StageCount, "下载 Fabric API",
                $"正在下载 {FabricApiDisplayName}…", 0, 0, 0, 0, 0, 0));

            var versions = await ModrinthVersionApi.GetVersionsAsync(
                    FabricApiProjectId,
                    gameVersions: [gameVersion],
                    loaders: ["fabric"],
                    cancellationToken: cancellationToken)
                .ConfigureAwait(false);

            var latest = versions.FirstOrDefault(
                v => v.PrimaryFile is { } f && !string.IsNullOrWhiteSpace(f.Url));
            if (latest?.PrimaryFile is null ||
                string.IsNullOrWhiteSpace(latest.PrimaryFile.Url))
            {
                Publish(new GameDownloadSnapshot(
                    NextRevision(), taskId, GameDownloadPhase.Downloading, instanceName,
                    MinecraftVersionInstaller.StageCount, "下载 Fabric API",
                    $"{FabricApiDisplayName} 无可用版本，已跳过。", 0, 0, 0, 0, 0, 0));
                return;
            }

            var targetPath = Path.Combine(modsDir, latest.PrimaryFile.Filename);
            if (File.Exists(targetPath))
                return;

            await ModDownloadService.DownloadAsync(
                    latest.PrimaryFile.Url, targetPath, cancellationToken: cancellationToken)
                .ConfigureAwait(false);

            Publish(new GameDownloadSnapshot(
                NextRevision(), taskId, GameDownloadPhase.Downloading, instanceName,
                MinecraftVersionInstaller.StageCount, "下载 Fabric API",
                $"{FabricApiDisplayName} 已下载至 mods/。", 0, 0, 0, 0, 0, 0));
        }
        catch (Exception exception)
        {
            // Fabric API 下载失败不应让整个实例安装失败，仅给出警告提示。
            Publish(new GameDownloadSnapshot(
                NextRevision(), taskId, GameDownloadPhase.Downloading, instanceName,
                MinecraftVersionInstaller.StageCount, "下载 Fabric API",
                $"{FabricApiDisplayName} 下载失败，请稍后手动安装：{exception.Message}",
                0, 0, 0, 0, 0, 0));
        }
    }

    private void PublishTerminal(
        long taskId,
        GameDownloadPhase phase,
        string stageName,
        string detail)
    {
        var previous = Current;
        if (previous.TaskId != taskId)
            return;
        Publish(previous with
        {
            Revision = NextRevision(),
            Phase = phase,
            StageName = stageName,
            Detail = detail
        });
    }

    private long NextRevision() => Interlocked.Increment(ref _revision);

    private void Publish(GameDownloadSnapshot snapshot)
    {
        lock (_gate)
            _current = snapshot;

        var handlers = Changed;
        if (handlers is null)
            return;
        foreach (Action<GameDownloadSnapshot> subscriber in handlers.GetInvocationList())
        {
            try
            {
                subscriber(snapshot);
            }
            catch (Exception exception)
            {
                System.Diagnostics.Debug.WriteLine(
                    $"GameDownloadService.Changed 订阅者异常：{exception}");
            }
        }
    }

    private sealed class InlineProgress<T>(Action<T> report) : IProgress<T>
    {
        public void Report(T value) => report(value);
    }
}

File: internal/download/game_download_service.go
package download

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"sync/atomic"

	"nyalauncher/internal/download/modrinth"
	"nyalauncher/internal/models"
)

// GameDownloadPhase 下载任务阶段。
type GameDownloadPhase int

const (
	GameDownloadIdle GameDownloadPhase = iota
	GameDownloadPreparing
	GameDownloadDownloading
	GameDownloadCompleted
	GameDownloadFailed
	GameDownloadCancelled
)

// GameDownloadSnapshot 下载任务快照（对外发布下载进度的不可变状态）。
// Go 无 record with 语法：使用 Clone + 字段赋值（见 snapshotWith）。
type GameDownloadSnapshot struct {
	Revision       int64
	TaskID         int64
	Phase          GameDownloadPhase
	VersionID      string
	StageIndex     int
	StageName      string
	Detail         string
	Percentage     float64
	CompletedBytes int64
	TotalBytes     int64
	CompletedFiles int
	TotalFiles     int
	BytesPerSecond float64
}

// IdleGameDownloadSnapshot 空闲快照。
func IdleGameDownloadSnapshot() GameDownloadSnapshot {
	return GameDownloadSnapshot{
		Phase:     GameDownloadIdle,
		StageName: "尚无下载任务",
		Detail:    "请在资源下载页选择 Minecraft 版本。",
	}
}

// HasTask 是否存在任务。
func (s GameDownloadSnapshot) HasTask() bool { return s.Phase != GameDownloadIdle }

// IsActive 任务是否活跃。
func (s GameDownloadSnapshot) IsActive() bool {
	return s.Phase == GameDownloadPreparing || s.Phase == GameDownloadDownloading
}

// IsTerminal 任务是否已终止。
func (s GameDownloadSnapshot) IsTerminal() bool {
	return s.Phase == GameDownloadCompleted || s.Phase == GameDownloadFailed || s.Phase == GameDownloadCancelled
}

// clone 返回快照副本（用于 "with" 语义的字段修改）。
func (s GameDownloadSnapshot) clone() GameDownloadSnapshot { return s }

// GameDownloadService 包装 MinecraftVersionInstaller，通过阶段/快照状态机对外发布下载进度。
type GameDownloadService struct {
	// OnChanged 快照变化回调（对应 C# Changed 事件；单订阅，
	// 多订阅方需自行分发——Wails 侧一般只桥接 EventsEmit 一个出口）。
	OnChanged func(GameDownloadSnapshot)

	gate       sync.Mutex
	installer  MinecraftVersionInstaller
	modLoader  ModLoaderInstaller
	current    GameDownloadSnapshot
	activeTask context.CancelFunc
	hasActive  bool
	revision   atomic.Int64
	taskID     atomic.Int64
}

// StageNames 下载阶段名（阶段 1-7）。
var GameDownloadStageNames = []string{
	"获取版本元数据",
	"分析下载清单",
	"下载游戏客户端",
	"下载依赖库",
	"下载资源索引",
	"下载游戏资源",
	"完成校验与安装",
}

// Current 当前快照。
func (s *GameDownloadService) Current() GameDownloadSnapshot {
	s.gate.Lock()
	defer s.gate.Unlock()
	return s.current
}

func (s *GameDownloadService) init() {
	s.gate.Lock()
	if s.current.Phase == GameDownloadIdle && s.current.StageName == "" {
		s.current = IdleGameDownloadSnapshot()
	}
	s.gate.Unlock()
}

// Start 下载并安装原版 Minecraft 版本。
func (s *GameDownloadService) Start(ctx context.Context, version models.MinecraftVersion) bool {
	if strings.TrimSpace(version.ID) == "" || strings.TrimSpace(version.URL) == "" {
		return false
	}
	return s.runDownloadTask(ctx, version.ID,
		fmt.Sprintf("Minecraft %s", version.ID),
		fmt.Sprintf("Minecraft %s 下载并安装完成", version.ID),
		version.ID,
		func(targetRoot string, progress InstallProgressFunc, taskCtx context.Context) error {
			return s.installer.Install(taskCtx, version.ID, version.URL, targetRoot, progress)
		},
		nil)
}

// StartModLoader 以 Mod Loader 模式下载：先确保原版已安装，再叠加安装 Loader。
func (s *GameDownloadService) StartModLoader(
	ctx context.Context,
	version models.MinecraftVersion,
	loader ModLoaderVersion,
	instanceName string,
	skipFabricApi bool,
) bool {
	if strings.TrimSpace(instanceName) == "" {
		return false
	}
	displayName := fmt.Sprintf("%s %s (Minecraft %s)", loader.Type, loader.LoaderVersion, version.ID)
	return s.runDownloadTask(ctx, instanceName, displayName,
		fmt.Sprintf("%s 安装完成", displayName), instanceName,
		func(targetRoot string, progress InstallProgressFunc, taskCtx context.Context) error {
			return s.modLoader.Install(taskCtx, loader, instanceName, targetRoot, version.ID, progress)
		},
		func(targetRoot, sourcePath string, taskID int64, taskCtx context.Context) error {
			if skipFabricApi || loader.Type != ModLoaderFabric {
				return nil
			}
			return s.downloadFabricAPIIfNeeded(taskCtx, loader, instanceName, targetRoot, sourcePath, version.ID, taskID)
		})
}

// runDownloadTask 下载任务的公共骨架：占位活跃任务槽 → 发布"准备中" → 挂接进度转发 →
// 执行安装 → 隔离目录骨架 / 活跃目录 / 实例刷新与选中 / 终态发布 → 任务槽清理。
// 原版安装与 Loader 安装除安装动作与 Fabric API 附加步骤外完全一致，
// 抽取为一个实现避免双份流程修复时漂移。
func (s *GameDownloadService) runDownloadTask(
	ctx context.Context,
	instanceID, displayName, completedDetail, selectID string,
	install func(targetRoot string, progress InstallProgressFunc, taskCtx context.Context) error,
	afterInstall func(targetRoot, sourcePath string, taskID int64, taskCtx context.Context) error,
) bool {
	s.init()
	s.gate.Lock()
	if s.hasActive {
		s.gate.Unlock()
		return false
	}
	taskCtx, cancel := context.WithCancel(ctx)
	s.activeTask = cancel
	s.hasActive = true
	taskID := s.taskID.Add(1)
	s.gate.Unlock()

	s.publish(GameDownloadSnapshot{
		Revision:   s.nextRevision(),
		TaskID:     taskID,
		Phase:      GameDownloadPreparing,
		VersionID:  instanceID,
		StageIndex: 1,
		StageName:  GameDownloadStageNames[0],
		Detail:     fmt.Sprintf("正在准备下载 %s", displayName),
	})

	defer func() {
		s.gate.Lock()
		s.activeTask = nil
		s.hasActive = false
		s.gate.Unlock()
		// 任务终态后自动解除全局暂停：暂停是任务的 UI 状态，
		// 不能泄漏到下一次下载（未暂停时 Resume 为无副作用空操作）
		DownloadPauseGate.Resume()
	}()

	targetRoot := s.resolveTargetMinecraftDirectory()
	progress := func(update MinecraftInstallProgress) {
		if taskCtx.Err() != nil || taskID != s.taskID.Load() {
			return
		}
		s.publish(GameDownloadSnapshot{
			Revision:       s.nextRevision(),
			TaskID:         taskID,
			Phase:          GameDownloadDownloading,
			VersionID:      instanceID,
			StageIndex:     update.StageIndex,
			StageName:      update.StageName,
			Detail:         update.Detail,
			Percentage:     update.Percentage(),
			CompletedBytes: update.CompletedBytes,
			TotalBytes:     update.TotalBytes,
			CompletedFiles: update.CompletedFiles,
			TotalFiles:     update.TotalFiles,
			BytesPerSecond: update.BytesPerSecond,
		})
	}

	if err := install(targetRoot, progress, taskCtx); err != nil {
		if taskCtx.Err() != nil || ctx.Err() != nil {
			s.publishTerminal(taskID, GameDownloadCancelled, "下载已取消", displayName+" 下载已取消")
			return false
		}
		s.publishTerminal(taskID, GameDownloadFailed, "下载失败", err.Error())
		return false
	}

	// 全局默认隔离开启时，为新版本预建隔离内容目录骨架。
	if ConfigDefaultVersionIsolation() {
		scaffoldIsolatedContentDirectory(targetRoot, instanceID)
	}

	sourcePath := ConfigGameDirectory()
	if strings.TrimSpace(sourcePath) == "" {
		ConfigSaveGameDirectory(targetRoot)
		sourcePath = targetRoot
	}

	if afterInstall != nil {
		if err := afterInstall(targetRoot, sourcePath, taskID, taskCtx); err != nil {
			if taskCtx.Err() != nil || ctx.Err() != nil {
				s.publishTerminal(taskID, GameDownloadCancelled, "下载已取消", displayName+" 下载已取消")
				return false
			}
			s.publishTerminal(taskID, GameDownloadFailed, "下载失败", err.Error())
			return false
		}
	}

	_ = RefreshInstances(sourcePath)
	SelectInstance(selectID)

	previous := s.Current()
	completed := previous.clone()
	completed.Revision = s.nextRevision()
	completed.Phase = GameDownloadCompleted
	completed.StageIndex = StageCount
	completed.StageName = GameDownloadStageNames[len(GameDownloadStageNames)-1]
	completed.Detail = completedDetail
	completed.Percentage = 100
	s.publish(completed)
	return true
}

// CancelActive 取消当前活跃下载任务。
func (s *GameDownloadService) CancelActive() bool {
	s.gate.Lock()
	defer s.gate.Unlock()
	if !s.hasActive || s.activeTask == nil {
		return false
	}
	s.activeTask()
	return true
}

// PauseActive 暂停当前活跃的下载任务：文件读取循环停在暂停门上，连接保持。
// 仅在任务活跃时生效；暂停详情会通过 OnChanged 发布。
func (s *GameDownloadService) PauseActive() bool {
	previous := s.Current()
	if !previous.IsActive() || DownloadPauseGate.IsPaused() {
		return false
	}

	DownloadPauseGate.Pause()
	updated := previous.clone()
	updated.Revision = s.nextRevision()
	updated.Detail = "已暂停（点击「继续下载」恢复）"
	s.publish(updated)
	return true
}

// ResumeActive 恢复被 PauseActive 暂停的全局下载。
func (s *GameDownloadService) ResumeActive() bool {
	if !DownloadPauseGate.IsPaused() {
		return false
	}

	previous := s.Current()
	DownloadPauseGate.Resume()
	// 恢复后安装器会在 ~120ms 内发布下一份进度，这里只刷新一次避免短暂显示过期详情
	if previous.IsActive() {
		updated := previous.clone()
		updated.Revision = s.nextRevision()
		updated.Detail = "已恢复下载"
		s.publish(updated)
	}
	return true
}

// resolveTargetMinecraftDirectory 解析下载目标根目录：
// 配置的游戏目录 → 环境变量 NYALAUNCHER_MINECRAFT_DIR → 默认 .minecraft；
// 若指向 versions/ 子目录则上提一级（与 C# 行为一致）。
func (s *GameDownloadService) resolveTargetMinecraftDirectory() string {
	configured := ConfigGameDirectory()
	if strings.TrimSpace(configured) == "" {
		configured = os.Getenv("NYALAUNCHER_MINECRAFT_DIR")
	}
	if strings.TrimSpace(configured) == "" {
		configured = EnsureDefaultMinecraftDirectory()
	}
	fullPath := filepath.Clean(configured)
	if strings.EqualFold(filepath.Base(fullPath), "versions") && filepath.Dir(fullPath) != fullPath {
		fullPath = filepath.Dir(fullPath)
	}

	_ = os.MkdirAll(fullPath, 0o755)
	_ = os.MkdirAll(filepath.Join(fullPath, "versions"), 0o755)
	return fullPath
}

// isolatedContentSubDirectories 版本隔离启用时的标准 Minecraft 内容子文件夹骨架。
// 游戏启动时会以该目录作为 --gameDir，提前创建可避免首次运行时目录缺失。
var isolatedContentSubDirectories = []string{
	"saves", "resourcepacks", "mods", "config",
	"crash-reports", "logs", "screenshots", "shaderpacks",
}

func scaffoldIsolatedContentDirectory(minecraftRoot, versionID string) {
	versionDirectory := filepath.Join(minecraftRoot, "versions", versionID)
	if info, err := os.Stat(versionDirectory); err != nil || !info.IsDir() {
		return
	}
	for _, sub := range isolatedContentSubDirectories {
		_ = os.MkdirAll(filepath.Join(versionDirectory, sub), 0o755)
	}
}

// fabricAPIProjectID Modrinth 上的 Fabric API 项目 ID（任何 Fabric 实例都需要的运行库）。
const fabricAPIProjectID = "P7dR8mSH"

const fabricAPIDisplayName = "Fabric API"

// downloadFabricAPIIfNeeded Fabric 实例安装完成后，默认把 Fabric API 一起下载到
// 该实例的 mods 目录。失败不阻断整个安装流程，仅作为警告提示。
func (s *GameDownloadService) downloadFabricAPIIfNeeded(
	ctx context.Context,
	loader ModLoaderVersion,
	instanceName, targetRoot, sourcePath, gameVersion string,
	taskID int64,
) error {
	if loader.Type != ModLoaderFabric {
		return nil
	}

	// 解析该实例实际的内容目录（版本隔离 / 共享布局 / 外部实例均可正确命中）。
	// 全局默认只作兜底传入，不覆盖自动检测（与 GameVersionIsolation.Resolve 语义一致）。
	var layoutContentDirectory string
	if ResolveInstanceLayoutHook != nil {
		layoutContentDirectory = ResolveInstanceLayoutHook(
			targetRoot, sourcePath, instanceName, gameVersion, ConfigDefaultVersionIsolation())
	}
	if strings.TrimSpace(layoutContentDirectory) == "" {
		layoutContentDirectory = targetRoot
	}
	modsDir := filepath.Join(layoutContentDirectory, "mods")
	if err := os.MkdirAll(modsDir, 0o755); err != nil {
		return err
	}

	publishStage := func(detail string) {
		s.publish(GameDownloadSnapshot{
			Revision:   s.nextRevision(),
			TaskID:     taskID,
			Phase:      GameDownloadDownloading,
			VersionID:  instanceName,
			StageIndex: StageCount,
			StageName:  "下载 Fabric API",
			Detail:     detail,
		})
	}

	publishStage(fmt.Sprintf("正在下载 %s…", fabricAPIDisplayName))

	versions, err := modrinthGetVersionsForLoader(ctx, fabricAPIProjectID, gameVersion)
	if err != nil {
		if ctx.Err() != nil {
			// 用户取消：不能伪装成"下载失败仅警告"，否则取消后任务仍按完成收尾
			return ctx.Err()
		}
		publishStage(fmt.Sprintf("%s 下载失败，请稍后手动安装：%v", fabricAPIDisplayName, err))
		return nil
	}

	var latest *fabricAPIFile
	for i := range versions {
		if primary := versions[i].PrimaryFile(); primary != nil && strings.TrimSpace(primary.URL) != "" {
			latest = &fabricAPIFile{filename: primary.Filename, url: primary.URL}
			break
		}
	}
	if latest == nil {
		publishStage(fmt.Sprintf("%s 无可用版本，已跳过。", fabricAPIDisplayName))
		return nil
	}

	targetPath := filepath.Join(modsDir, latest.filename)
	if _, err := os.Stat(targetPath); err == nil {
		return nil
	}

	if err := DownloadFileToPath(ctx, latest.url, targetPath, nil); err != nil {
		if ctx.Err() != nil {
			return ctx.Err()
		}
		publishStage(fmt.Sprintf("%s 下载失败，请稍后手动安装：%v", fabricAPIDisplayName, err))
		return nil
	}
	publishStage(fmt.Sprintf("%s 已下载至 mods/。", fabricAPIDisplayName))
	return nil
}

type fabricAPIFile struct {
	filename string
	url      string
}

func (s *GameDownloadService) publishTerminal(taskID int64, phase GameDownloadPhase, stageName, detail string) {
	previous := s.Current()
	if previous.TaskID != taskID {
		return
	}
	updated := previous.clone()
	updated.Revision = s.nextRevision()
	updated.Phase = phase
	updated.StageName = stageName
	updated.Detail = detail
	s.publish(updated)
}

func (s *GameDownloadService) nextRevision() int64 { return s.revision.Add(1) }

// publish 发布快照。快照可能来自多个并行下载线程：序号必须在锁内分配并写入，
// 否则线程 A 拿到旧序号却后进锁，会把新快照覆盖回旧状态。
func (s *GameDownloadService) publish(snapshot GameDownloadSnapshot) {
	if snapshot.Revision == 0 {
		snapshot.Revision = s.nextRevision()
	}
	s.gate.Lock()
	if snapshot.Revision < s.current.Revision {
		s.gate.Unlock()
		return
	}
	s.current = snapshot
	handler := s.OnChanged
	s.gate.Unlock()

	if handler != nil {
		// 订阅者异常不影响发布方（对应 C# 逐订阅者 try/catch）
		func() {
			defer func() { _ = recover() }()
			handler(snapshot)
		}()
	}
}

// modrinthGetVersionsForLoader 查询 Fabric API 在指定 MC 版本 + fabric loader 下的版本列表。
func modrinthGetVersionsForLoader(ctx context.Context, projectID, gameVersion string) ([]models.ModrinthVersion, error) {
	return modrinth.GetVersionsForCombo(ctx, projectID, gameVersion, "fabric")
}
